In [1]:
import geopandas as gpd
import numpy as np
import pandas as pd
import matplotlib as plt
from shapely.validation import make_valid


## Philippine River Dataset 

This dataset contains all types of waterways located in the Philippines. For this assessment we will only be utilizing waterway specifically `river` and `stream`. 

In [ ]:
# Load dataset
river_df = gpd.read_file("hotosm_phl_waterways_lines_shp.shp")
river_df


name name_en waterway covered width depth layer blockage  \
0           Cagayan River    None    river    None  None  None  None     None   
1                    None    None    river    None  None  None  None     None   
2                    None    None   stream    None  None  None  None     None   
3                    None    None   stream    None  None  None    -1     None   
4                    None    None   stream    None  None  None  None     None   
...                   ...     ...      ...     ...   ...   ...   ...      ...   
121214       Angono River    None    river    None  None  None  None     None   
121215  Pamoantogbo River    None    river    None  None  None  None     None   
121216     Odiongan River    None    river    None  None  None  None     None   
121217     Mandulog River    None    river    None  None  None  None     None   
121218        Gamay River    None    river    None  None  None  None     None   

         tunnel natural water   source name_fil      osm_id   osm_type  \
0          None    None  None  landsat     None   909866165  ways_line   
1          None    None  None     None     None   245941601  ways_line   
2          None    None  None     None     None   502889972  ways_line   
3       culvert    None  None     None     None  1184195783  ways_line   
4          None    None  None     None     None  1317786873  ways_line   
...         ...     ...   ...      ...      ...         ...        ...   
121214     None    None  None     None     None    15246527  relations   
121215     None    None  None     None     None    13036436  relations   
121216     None    None  None     None     None    12720902  relations   
121217     None    None  None     None     None    12715038  relations   
121218     None    None  None     None     None    12273033  relations   

                                                 geometry  
0       LINESTRING (121.34927 15.99296, 121.34949 15.9...  
1       LINESTRING (124.87041 11.16783, 124.87038 11.1...  
2       LINESTRING (124.97525 10.88445, 124.97532 10.8...  
3        LINESTRING (125.51376 7.05871, 125.51389 7.0586)  
4       LINESTRING (123.39114 12.27959, 123.39137 12.2...  
...                                                   ...  
121214  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
121215  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
121216  LINESTRING (125.19814 8.82519, 125.19764 8.825...  
121217  LINESTRING (124.24519 8.25548, 124.24604 8.255...  
121218  LINESTRING (125.23851 12.41278, 125.23868 12.4...  

[121219 rows x 16 columns]

In [3]:
river_final = river_df[river_df["waterway"].isin(["river","stream"])]
river_final = river_final[["name", "waterway", "geometry"]]

river_final

name waterway  \
0           Cagayan River    river   
1                    None    river   
2                    None   stream   
3                    None   stream   
4                    None   stream   
...                   ...      ...   
121214       Angono River    river   
121215  Pamoantogbo River    river   
121216     Odiongan River    river   
121217     Mandulog River    river   
121218        Gamay River    river   

                                                 geometry  
0       LINESTRING (121.34927 15.99296, 121.34949 15.9...  
1       LINESTRING (124.87041 11.16783, 124.87038 11.1...  
2       LINESTRING (124.97525 10.88445, 124.97532 10.8...  
3        LINESTRING (125.51376 7.05871, 125.51389 7.0586)  
4       LINESTRING (123.39114 12.27959, 123.39137 12.2...  
...                                                   ...  
121214  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
121215  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
121216  LINESTRING (125.19814 8.82519, 125.19764 8.825...  
121217  LINESTRING (124.24519 8.25548, 124.24604 8.255...  
121218  LINESTRING (125.23851 12.41278, 125.23868 12.4...  

[97950 rows x 3 columns]

### Exploratory Data Analysis 

In [ ]:
# ensure that all values for geometry is in the right format 
if river_final.crs is None or river_final.crs.to_epsg() != 4326:
    river_final = river_final.to_crs(4326)

In [5]:
river_final.isnull().sum()

name        88416
waterway        0
geometry        0
dtype: int64

In [6]:
river_final.duplicated().sum()


np.int64(6)

In [7]:
river_final[river_final.duplicated()]


name waterway  \
121111    Lubogan River    river   
121114    Dalanas River    river   
121115    Litoban River    river   
121170  Bitanagan River    river   
121212     Anayan River    river   
121216   Odiongan River    river   

                                                 geometry  
121111  LINESTRING (125.42837 7.05983, 125.42845 7.059...  
121114  LINESTRING (122.16361 11.37186, 122.16361 11.3...  
121115  LINESTRING (122.2797 7.68224, 122.27956 7.6822...  
121170  LINESTRING (126.19383 7.07723, 126.19354 7.077...  
121212  LINESTRING (120.85559 17.83885, 120.85547 17.8...  
121216  LINESTRING (125.19814 8.82519, 125.19764 8.825...

In [8]:
river_cleaned_df = river_final.dropna()
river_cleaned_df = river_cleaned_df.drop_duplicates()

In [9]:
river_cleaned_df

name waterway  \
0           Cagayan River    river   
7             Lopoy Creek   stream   
23          Tuganay River    river   
47           Matiao River    river   
56        Calumpang River    river   
...                   ...      ...   
121213       Bugsuk River    river   
121214       Angono River    river   
121215  Pamoantogbo River    river   
121217     Mandulog River    river   
121218        Gamay River    river   

                                                 geometry  
0       LINESTRING (121.34927 15.99296, 121.34949 15.9...  
7       LINESTRING (125.41293 7.06458, 125.41301 7.064...  
23      LINESTRING (125.51876 7.56096, 125.51912 7.560...  
47      LINESTRING (126.07515 7.23193, 126.07495 7.232...  
56      LINESTRING (121.07517 13.76744, 121.07496 13.7...  
...                                                   ...  
121213  LINESTRING (117.29818 8.29036, 117.29876 8.291...  
121214  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
121215  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
121217  LINESTRING (124.24519 8.25548, 124.24604 8.255...  
121218  LINESTRING (125.23851 12.41278, 125.23868 12.4...  

[9528 rows x 3 columns]

## OCHA Dataset

In [ ]:
# Load Dataset
adm2 = gpd.read_file("phl_adm_psa_namria_20231106_shp/phl_admbnda_adm2_psa_namria_20231106.shp").to_crs(4326)
adm3 = gpd.read_file("phl_adm_psa_namria_20231106_shp/phl_admbnda_adm3_psa_namria_20231106.shp").to_crs(4326)

Index(['ADM2_EN', 'ADM2_PCODE', 'ADM1_EN', 'ADM1_PCODE', 'ADM0_EN',
       'ADM0_PCODE'],
      dtype='object')
Index(['ADM3_EN', 'ADM3_PCODE', 'ADM2_EN', 'ADM2_PCODE', 'ADM1_EN',
       'ADM1_PCODE'],
      dtype='object')


In [ ]:
rivers = river_cleaned_df.copy()

# Combine provinces
rivers = gpd.sjoin(rivers, adm2[["ADM1_EN","ADM2_EN","geometry"]],
                   how="left", predicate="intersects")

# Combine cities and municipalities  
rivers = gpd.sjoin(rivers, adm3[["ADM3_EN","ADM2_EN","geometry"]],
                   how="left", predicate="intersects",
                   lsuffix="prov", rsuffix="mun")


In [12]:
river_city = rivers[[
    "name", "waterway",
    "ADM1_EN", "ADM2_EN_prov", "ADM3_EN",
    "geometry"
]].reset_index(drop=True)


In [13]:
river_city 

name waterway                          ADM1_EN  \
0          Cagayan River    river       Region II (Cagayan Valley)   
1            Lopoy Creek   stream         Region XI (Davao Region)   
2          Tuganay River    river         Region XI (Davao Region)   
3          Tuganay River    river         Region XI (Davao Region)   
4          Tuganay River    river         Region XI (Davao Region)   
...                  ...      ...                              ...   
13818       Angono River    river         Region IV-A (Calabarzon)   
13819  Pamoantogbo River    river  Region IX (Zamboanga Peninsula)   
13820     Mandulog River    river     Region X (Northern Mindanao)   
13821        Gamay River    river    Region VIII (Eastern Visayas)   
13822        Gamay River    river    Region VIII (Eastern Visayas)   

            ADM2_EN_prov            ADM3_EN  \
0          Nueva Vizcaya  Alfonso Castaneda   
1          Davao del Sur         Davao City   
2          Davao del Sur         Davao City   
3          Davao del Sur        Santo Tomas   
4        Davao del Norte         Davao City   
...                  ...                ...   
13818              Rizal             Angono   
13819  Zamboanga Sibugay              Payao   
13820    Lanao del Norte        Iligan City   
13821     Northern Samar              Gamay   
13822     Northern Samar            Mapanas   

                                                geometry  
0      LINESTRING (121.34927 15.99296, 121.34949 15.9...  
1      LINESTRING (125.41293 7.06458, 125.41301 7.064...  
2      LINESTRING (125.51876 7.56096, 125.51912 7.560...  
3      LINESTRING (125.51876 7.56096, 125.51912 7.560...  
4      LINESTRING (125.51876 7.56096, 125.51912 7.560...  
...                                                  ...  
13818  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
13819  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
13820  LINESTRING (124.24519 8.25548, 124.24604 8.255...  
13821  LINESTRING (125.23851 12.41278, 125.23868 12.4...  
13822  LINESTRING (125.23851 12.41278, 125.23868 12.4...  

[13823 rows x 6 columns]

Notice that our rows increased from <b> 9,528 </b> to <b> 13,823 </b>. We treat this increase not as an error but because it means that some rivers flows across multiple provinces/city. 

In [31]:
df = river_city.copy()
df = df[df["waterway"].isin(["river","stream"])].reset_index(drop=True)
if df.crs is None or df.crs.to_epsg() != 4326:
    df = df.to_crs(4326)

# Rename columns for clarity
rename_map = {
    "name": "Name",
    "waterway": "Type",           
    "ADM1_EN": "Region",
    "ADM2_EN_prov": "Province",
    "ADM3_EN": "City_Muni"
}
df = df.rename(columns=rename_map)

# Representative coordinates
rep = df.geometry.representative_point()
df["Lon"] = rep.x.round(6)
df["Lat"] = rep.y.round(6)

# Fix text structure 
for col in ["Name","Region","Province","City_Muni"]:
    if col in df.columns:
        df[col] = (
            df[col]
            .astype(str)
            .str.strip()
            .replace({"nan": np.nan})
        )

df["Type"] = df["Type"].str.lower()

# Since some data were given a name tag we normalize it 
df["Name"] = df["Name"].replace(
    to_replace=r"^\d{1,2}/\d{1,2}/\d{4}.*", 
    value="Unnamed Stream (Digitized 2013)",
    regex=True
)
df["Name"] = df["Name"].replace(
    to_replace=r"^(source of drinking water\?|toStream|to be named|stream along.*)$",
    value="Unnamed Stream",
    regex=True
)

# Consistent Title 
df["Name"] = df["Name"].str.title()

# Drop duplicates  
dedupe_keys = ["Name","Type","Province","City_Muni"]
df = df.sort_values(dedupe_keys + ["Region"]).drop_duplicates(subset=dedupe_keys, keep="first")

# Order columns
cols = ["Name","Type","Region","Province","City_Muni","Lat","Lon"]
df_out = df[cols].reset_index(drop=True)

# CSV Export 
out_path = "River_dataset.csv"
df_out.to_csv(out_path, index=False, encoding="utf-8")
out_path


'River_dataset.csv'

# Supplementary Data

## River Catchment Characteristics Dataset

This dataset includes <b> catchment polygons, stream networks, rainfall/runoff attributes, and mean flow statistics </b> which allows us to identify <b> hydrological basics feeding each river, </b> and estimate water discharge. By using this data we are able to get an estimate on its possible usage for run-of-river projects. 

In [ ]:
TOL_M_NEAR = 200   # consider a line "on_n128" if <= this distance (m)
TOL_M_FBK  = 500   # fallback tolerance to nearest catchment if rep-point misses (m)


# Using the previous dataset 
rivers = river_city[river_city["waterway"].isin(["river", "stream"])].copy()
if rivers.crs is None or rivers.crs.to_epsg() != 4326:
    rivers = rivers.to_crs(4326)
expected_rows = len(rivers)

# Consolidate n128 catchments & stream network
catch_river  = gpd.read_file(
    "characteristics/Philippines_GIS_catchments_n128/philippines_gis_catchments_n128.shp"
).to_crs(4326)

catch_stream = gpd.read_file(
    "characteristics/Philippines_GIS_stream_network_n128/philippines_gis_stream_network_n128.shp"
).to_crs(4326)

# Stable key + representative point
rivers = rivers.reset_index(drop=True)
rivers["row_id"] = rivers.index
rivers["rep_pt"] = rivers.geometry.representative_point()
pts = gpd.GeoDataFrame(rivers.drop(columns="geometry"), geometry="rep_pt", crs=4326)

# Keep only known catchment fields 
catch_keep = [c for c in [
    "New_Name","Catch_area","Catch_elev","Catch_slp","Catch_hyp","Drainage_D","Catch_rel","geometry"
] if c in catch_river.columns]

# Dissolve by New_Name to remove internal overlaps/multiparts (prevents 1→many) this ensures that a river won't be duplicated 
catch_base = catch_river[catch_keep].copy()
if "New_Name" in catch_base.columns:
    catch_base = catch_base.dissolve(by="New_Name", as_index=False)

# Primary: point-in-polygon (rep-point within catchment)
join1 = gpd.sjoin(pts, catch_base, how="left", predicate="within")

# If sjoin brings index_right (match id), use polygon area to keep dominant match per row_id
if "index_right" in join1.columns:
    poly_area = catch_base.to_crs(3857).area
    area_tbl  = (catch_base.reset_index()
                          .rename(columns={"index":"index_right"})
                          .assign(poly_area=poly_area.values)[["index_right","poly_area"]])
    join1 = join1.merge(area_tbl, on="index_right", how="left")
    join1 = (join1.sort_values(["row_id","poly_area"], ascending=[True, False])
                   .drop_duplicates("row_id"))
    join1 = join1.drop(columns=[c for c in ["index_right","poly_area"] if c in join1.columns])
else:
    join1 = join1.sort_values("row_id").drop_duplicates("row_id")

# Normalize name
if "New_Name" in join1.columns:
    join1 = join1.rename(columns={"New_Name":"Catchment"})

# Fallback: nearest polygon for any misses (align by row_id)
miss_mask = join1["Catchment"].isna() if "Catchment" in join1.columns else pd.Series(False, index=join1.index)
if miss_mask.any():
    miss_pts = rivers.loc[miss_mask, ["row_id","rep_pt"]].copy()
    miss_pts = gpd.GeoDataFrame(miss_pts.set_geometry("rep_pt"), geometry="rep_pt", crs=4326).to_crs(3857)

    polys_m  = catch_base.to_crs(3857)
    fb = gpd.sjoin_nearest(miss_pts, polys_m, how="left", distance_col="d_near")
    fb = fb[fb["d_near"] <= TOL_M_FBK].to_crs(4326)

    if "New_Name" in fb.columns:
        fb = fb.rename(columns={"New_Name":"Catchment"})

    assign_cols = [c for c in ["Catchment","Catch_area","Catch_elev","Catch_slp","Catch_hyp","Drainage_D","Catch_rel"] if c in fb.columns]
    fb_small = fb[["row_id"] + assign_cols].drop_duplicates("row_id")

    join1 = join1.merge(fb_small.add_suffix("_fb"), left_on="row_id", right_on="row_id_fb", how="left")
    for col in assign_cols:
        join1[col] = join1[col].fillna(join1[f"{col}_fb"])
    drop_helpers = ["row_id_fb"] + [f"{c}_fb" for c in assign_cols]
    join1 = join1.drop(columns=[c for c in drop_helpers if c in join1.columns])

# Restore original line geometry
if "rep_pt" in join1.columns:
    join1 = join1.drop(columns=["rep_pt"])
rivers_catch = gpd.GeoDataFrame(join1, geometry=rivers.geometry, crs=4326)

# Enforce 1:1
rivers_catch = rivers_catch.sort_values("row_id").drop_duplicates("row_id")
assert len(rivers_catch) == expected_rows, f"[Catchment step] Expected {expected_rows}, got {len(rivers_catch)}"

# Try to keep an order/strahler column + an id if present
order_candidates = [c for c in catch_stream.columns if c.lower() in {"order","strahler","streamord"}]
id_candidates    = [c for c in catch_stream.columns if c.lower() in {"stream_id","river_id","rid","fid","id"}]
keep_cols = (order_candidates[:1] + id_candidates[:1])

cs_use = catch_stream[keep_cols + ["geometry"]] if keep_cols else catch_stream[["geometry"]].copy()

riv_m = rivers_catch.to_crs(3857)
cs_m  = cs_use.to_crs(3857)
near  = gpd.sjoin_nearest(riv_m, cs_m, how="left", distance_col="dist_m").to_crs(4326)

# Keep only the single nearest per row_id
near = near.sort_values(["row_id","dist_m"]).drop_duplicates("row_id")

# Flag if aligned to the modeled network
near["on_n128"] = near["dist_m"].notna() & (near["dist_m"] <= TOL_M_NEAR)
if keep_cols:
    near.loc[~near["on_n128"], keep_cols] = np.nan

assert len(near) == expected_rows, f"[n128 step] Expected {expected_rows}, got {len(near)}"

# Compute Hydro Score
for c in ["Catch_area","Catch_elev","Catch_slp","Catch_rel"]:
    if c in near.columns:
        near[c] = pd.to_numeric(near[c], errors="coerce")

def hydro_rule_typeaware(row):
    if row["waterway"] == "stream":
        area_min, slope_min, relief_min, elev_min = 50, 8, 80, 120
    else:
        area_min, slope_min, relief_min, elev_min = 100, 10, 100, 150
    return (
        (row.get("Catch_area", np.nan) >= area_min) and
        (row.get("Catch_slp",  np.nan) >= slope_min) and
        (row.get("Catch_rel",  np.nan) >= relief_min) and
        (row.get("Catch_elev", np.nan) >= elev_min)
    )

def hydro_score(row):
    a = np.clip((row.get("Catch_area", 0)-50)/(300-50),   0, 1)
    s = np.clip((row.get("Catch_slp",  0)-5) /(20-5),     0, 1)
    r = np.clip((row.get("Catch_rel",  0)-50)/(300-50),   0, 1)
    e = np.clip((row.get("Catch_elev", 0)-100)/(600-100), 0, 1)
    return round(100*(0.35*a + 0.25*r + 0.25*s + 0.15*e), 1)

near["hydro_usable"] = near.apply(lambda x: "Yes" if hydro_rule_typeaware(x) else "No", axis=1)
near["hydro_score"]  = near.apply(hydro_score, axis=1)

scored_river = [
    "row_id","name","waterway","ADM1_EN","ADM2_EN_prov","ADM3_EN",
    "Catchment","Catch_area","Catch_elev","Catch_slp","Catch_hyp","Drainage_D","Catch_rel",
    *keep_cols, "dist_m","on_n128",
    "hydro_usable","hydro_score","geometry"
]
scored_river = [c for c in scored_river if c in near.columns]
scored_river = near[scored_river].reset_index(drop=True)


In [15]:
scored_river

row_id               name waterway                          ADM1_EN  \
0           0      Cagayan River    river       Region II (Cagayan Valley)   
1           1        Lopoy Creek   stream         Region XI (Davao Region)   
2           2      Tuganay River    river         Region XI (Davao Region)   
3           3      Tuganay River    river         Region XI (Davao Region)   
4           4      Tuganay River    river         Region XI (Davao Region)   
...       ...                ...      ...                              ...   
13818   13818       Angono River    river         Region IV-A (Calabarzon)   
13819   13819  Pamoantogbo River    river  Region IX (Zamboanga Peninsula)   
13820   13820     Mandulog River    river     Region X (Northern Mindanao)   
13821   13821        Gamay River    river    Region VIII (Eastern Visayas)   
13822   13822        Gamay River    river    Region VIII (Eastern Visayas)   

            ADM2_EN_prov            ADM3_EN       Catchment  Catch_area  \
0          Nueva Vizcaya  Alfonso Castaneda         Cagayan   27684.068   
1          Davao del Sur         Davao City             NaN         NaN   
2          Davao del Sur         Davao City           Tagum    3167.920   
3          Davao del Sur        Santo Tomas           Tagum    3167.920   
4        Davao del Norte         Davao City           Tagum    3167.920   
...                  ...                ...             ...         ...   
13818              Rizal             Angono  Marikina_Pasig    3960.976   
13819  Zamboanga Sibugay              Payao             NaN         NaN   
13820    Lanao del Norte        Iligan City        Mandulog     776.031   
13821     Northern Samar              Gamay             NaN         NaN   
13822     Northern Samar            Mapanas             NaN         NaN   

       Catch_elev  Catch_slp  Catch_hyp  Drainage_D  Catch_rel  streamord  \
0         514.605     16.619      0.179       0.832   2871.037        3.0   
1             NaN        NaN        NaN         NaN        NaN        NaN   
2         213.310     15.334      0.140       0.951   1528.282        NaN   
3         213.310     15.334      0.140       0.951   1528.282        NaN   
4         213.310     15.334      0.140       0.951   1528.282        NaN   
...           ...        ...        ...         ...        ...        ...   
13818     155.505      7.225      0.072       0.976   2154.866        2.0   
13819         NaN        NaN        NaN         NaN        NaN        NaN   
13820     633.527     16.620      0.452       0.813   1400.139        5.0   
13821         NaN        NaN        NaN         NaN        NaN        NaN   
13822         NaN        NaN        NaN         NaN        NaN        NaN   

             dist_m  on_n128 hydro_usable  hydro_score  \
0         21.601520     True          Yes         91.8   
1       9092.042719    False           No          NaN   
2       3590.409163    False          Yes         80.6   
3       3590.409163    False          Yes         80.6   
4       3590.409163    False          Yes         80.6   
...             ...      ...          ...          ...   
13818      0.000000     True           No         65.4   
13819  12154.562002    False           No          NaN   
13820      0.000000     True          Yes         94.4   
13821   3999.619763    False           No          NaN   
13822   3999.619763    False           No          NaN   

                                                geometry  
0      LINESTRING (121.34927 15.99296, 121.34949 15.9...  
1      LINESTRING (125.41293 7.06458, 125.41301 7.064...  
2      LINESTRING (125.51876 7.56097, 125.51912 7.560...  
3      LINESTRING (125.51876 7.56097, 125.51912 7.560...  
4      LINESTRING (125.51876 7.56097, 125.51912 7.560...  
...                                                  ...  
13818  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
13819  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
13820  LINESTRIN

## Philippine National Hydrological Model Dataset

This dataset gives a score to each `river` and `stream` which can be used to determine how usable it is for a company to build a hydropowered dam. 

In [ ]:
# Load CSV file 
runoff = pd.read_csv("Runoff_all.csv")
runoff.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88 entries, 0 to 87
Columns: 512 entries, Unnamed: 0 to NAME_1
dtypes: float64(510), int64(1), object(1)
memory usage: 352.1+ KB


Unnamed: 0  RUNOFF1979.01.03  RUNOFF1979.02.03  RUNOFF1979.03.03  \
0           1             1.960              5.59             6.438   
1           2            14.350              5.81            49.590   
2           3            11.700             10.54            57.370   
3           4            15.189              6.17             6.617   
4           5             6.510              4.47             0.809   

   RUNOFF1979.04.03  RUNOFF1979.05.03  RUNOFF1979.06.03  RUNOFF1979.07.03  \
0             26.42            136.02            232.59            284.78   
1             32.68             57.06            100.27             89.77   
2             23.75             72.02            157.28            105.61   
3             44.04             73.79            203.39            154.74   
4            125.01             77.68             67.89            101.82   

   RUNOFF1979.08.03  RUNOFF1979.09.03  ...  RUNOFF_mean_1999-2018_7  \
0            197.32            168.13  ...                   318.50   
1             59.67             72.31  ...                    98.08   
2             99.53             87.00  ...                   114.35   
3            125.96            115.88  ...                   173.38   
4            190.97             73.51  ...                   149.41   

   RUNOFF_mean_1999-2018_8  RUNOFF_mean_1999-2018_9  RUNOFF_mean_1999-2018_10  \
0                   283.44                   214.80                    148.13   
1                    90.91                    88.30                     81.33   
2                   120.95                   128.43                    106.34   
3                   126.32                   148.92                    127.86   
4                   136.50                   140.98                     89.41   

   RUNOFF_mean_1999-2018_11  RUNOFF_mean_1999-2018_12  \
0                     59.11                     47.19   
1                     93.72                    102.65   
2                     88.67                     95.03   
3                    111.72                    171.19   
4                     55.89                     92.97   

   RUNOFF_mean_1999-2018_annual  RUNOFF_mean_1999-2018_wet  \
0                        122.32                     210.34   
1                         93.67                      85.45   
2                        103.66                     109.71   
3                        113.16                     128.66   
4                         80.23                     109.57   

   RUNOFF_mean_1999-2018_dry            NAME_1  
0                      34.31              Abra  
1                     101.89  Agusan del Norte  
2                      97.61    Agusan del Sur  
3                      97.66             Aklan  
4                      50.88             Albay  

[5 rows x 512 columns]

In [ ]:
def load_runoff(path="Runoff_all.csv"):
    for sep in [";", ",", "\t", "|"]:
        try:
            df = pd.read_csv(path, sep=sep, engine="python")
            if df.shape[1] > 1:
                df.columns = df.columns.str.strip()
                for c in df.columns:
                    if df[c].dtype == object:
                        df[c] = df[c].astype(str).str.strip()
                return df
        except Exception:
            pass
    return pd.read_csv(path)

runoff = load_runoff("Runoff_all.csv")

# Detect columns (province + runoff value)
prov_col = next((c for c in runoff.columns
                 if str(c).upper() in {"NAME_1","PROVINCE","ADM2","ADM2_EN","PROVINCENAME"}), None)
val_col  = next((c for c in runoff.columns if "runoff" in str(c).lower()), None)
if val_col is None:
    num_cols = [c for c in runoff.columns if pd.api.types.is_numeric_dtype(runoff[c])]
    val_col  = num_cols[0] if num_cols else None
assert prov_col and val_col, f"Runoff CSV: could not detect columns. Found: {list(runoff.columns)}"

# Province normalization and mapping
def norm(s: pd.Series) -> pd.Series:
    return (s.astype(str)
              .str.normalize("NFKD")
              .str.encode("ascii", errors="ignore").str.decode("ascii")
              .str.upper()
              .str.replace(r"[^A-Z0-9]", "", regex=True))

PROV_MAP = {
    "NCR":"METROMANILA","NATIONALCAPITALREGION":"METROMANILA","METROPOLITANMANILA":"METROMANILA",
    "DAVAO DEL SUR".upper().replace(" ",""):"DAVAODELSUR",
    "DAVAO DEL NORTE".upper().replace(" ",""):"DAVAODELNORTE",
    "COTABATONORTH":"NORTHCOTABATO","COTABATOSOUTH":"SOUTHCOTABATO",
    "MAGUINDANAODELNORTE":"MAGUINDANAODELNORTE","MAGUINDANAODELSUR":"MAGUINDANAODELSUR",
    "DINAGATISLANDS":"DINAGATISLANDS","LANAODELNORTE":"LANAODELNORTE","LANAODELSUR":"LANAODELSUR",
    "OCCIDENTALMINDORO":"OCCIDENTALMINDORO","ORIENTALMINDORO":"ORIENTALMINDORO",
    "DAVAOOCCIDENTAL":"DAVAOOCCIDENTAL","DAVAOORIENTAL":"DAVAOORIENTAL",
}

runoff = runoff.copy()
runoff["prov_key"] = norm(runoff[prov_col]).replace(PROV_MAP)

scored_river = scored_river.copy()
if "ADM2_EN_prov" not in scored_river.columns:
    raise ValueError("`scored_river` must contain 'ADM2_EN_prov' (province).")
scored_river["prov_key"] = norm(scored_river["ADM2_EN_prov"]).replace(PROV_MAP)

# Clean any prior scoring columns to avoid collisions
drop_cols = [c for c in scored_river.columns if c in
             {"Runoff_from_PNHMD","Runoff_mm_yr","Q_m3s","Power_MW","HPI","Final_Status","Final_Usable"}]
if drop_cols:
    scored_river = scored_river.drop(columns=drop_cols)

# Merge runoff 
right = runoff[["prov_key", val_col]].rename(columns={val_col:"Runoff_from_PNHMD"}).drop_duplicates("prov_key")
before_n = len(scored_river)
scored_river = scored_river.merge(right, on="prov_key", how="left", validate="m:1")
assert len(scored_river) == before_n, "Row count changed — merge should be m:1."

# Normalize units 
med_raw = pd.to_numeric(scored_river["Runoff_from_PNHMD"], errors="coerce").median()
if pd.notna(med_raw) and 0.05 <= med_raw <= 20:
    scored_river["Runoff_mm_yr"] = pd.to_numeric(scored_river["Runoff_from_PNHMD"], errors="coerce") * 1000.0
    runoff_unit = "m/yr → converted to mm/yr"
else:
    scored_river["Runoff_mm_yr"] = pd.to_numeric(scored_river["Runoff_from_PNHMD"], errors="coerce")
    runoff_unit = "mm/yr (as provided)"

mask_m = scored_river["Runoff_mm_yr"] < 50
scored_river.loc[mask_m, "Runoff_mm_yr"] *= 1000.0

# Compute flow, power
for c in ["Runoff_mm_yr","Catch_area","Catch_rel"]:
    if c in scored_river.columns:
        scored_river[c] = pd.to_numeric(scored_river[c], errors="coerce").clip(lower=0)

SEC_PER_YR = 31_536_000
RHO, G, ETA = 1000.0, 9.81, 0.75

scored_river["Q_m3s"]   = (scored_river["Runoff_mm_yr"] * scored_river["Catch_area"]) / (SEC_PER_YR * 1000.0)
scored_river["Power_MW"] = (RHO * G * scored_river["Q_m3s"] * scored_river["Catch_rel"] * ETA) / 1e6

# Compute for HPI 
def _nz(x): return 0.0 if pd.isna(x) else float(x)
def hpi_row(r):
    runoff_s = np.clip(_nz(r.get("Runoff_mm_yr"))/2000.0, 0, 1)
    relief_s = np.clip(_nz(r.get("Catch_rel"))   / 500.0, 0, 1)
    power_s  = np.clip(_nz(r.get("Power_MW"))    / 5.0, 0, 1)
    flow_s   = np.clip(_nz(r.get("Q_m3s"))       / 1.0, 0, 1)
    return round((0.30*runoff_s + 0.30*relief_s + 0.30*power_s + 0.10*flow_s)*100.0, 1)

scored_river["HPI"] = scored_river.apply(hpi_row, axis=1)

def hpi_to_status(x):
    if pd.isna(x): return "Insufficient Data"
    if x >= 80:    return "Excellent"
    if x >= 60:    return "Good"
    if x >= 40:    return "Moderate"
    return "Poor"

scored_river["Final_Status"] = scored_river["HPI"].apply(hpi_to_status)
scored_river["Final_Usable"] = scored_river["Final_Status"].isin(["Excellent","Good"])



In [25]:
scored_river

row_id               name waterway                          ADM1_EN  \
0           0      Cagayan River    river       Region II (Cagayan Valley)   
1           1        Lopoy Creek   stream         Region XI (Davao Region)   
2           2      Tuganay River    river         Region XI (Davao Region)   
3           3      Tuganay River    river         Region XI (Davao Region)   
4           4      Tuganay River    river         Region XI (Davao Region)   
...       ...                ...      ...                              ...   
13818   13818       Angono River    river         Region IV-A (Calabarzon)   
13819   13819  Pamoantogbo River    river  Region IX (Zamboanga Peninsula)   
13820   13820     Mandulog River    river     Region X (Northern Mindanao)   
13821   13821        Gamay River    river    Region VIII (Eastern Visayas)   
13822   13822        Gamay River    river    Region VIII (Eastern Visayas)   

            ADM2_EN_prov            ADM3_EN       Catchment  Catch_area  \
0          Nueva Vizcaya  Alfonso Castaneda         Cagayan   27684.068   
1          Davao del Sur         Davao City             NaN         NaN   
2          Davao del Sur         Davao City           Tagum    3167.920   
3          Davao del Sur        Santo Tomas           Tagum    3167.920   
4        Davao del Norte         Davao City           Tagum    3167.920   
...                  ...                ...             ...         ...   
13818              Rizal             Angono  Marikina_Pasig    3960.976   
13819  Zamboanga Sibugay              Payao             NaN         NaN   
13820    Lanao del Norte        Iligan City        Mandulog     776.031   
13821     Northern Samar              Gamay             NaN         NaN   
13822     Northern Samar            Mapanas             NaN         NaN   

       Catch_elev  Catch_slp  ...  hydro_score  \
0         514.605     16.619  ...         91.8   
1             NaN        NaN  ...          NaN   
2         213.310     15.334  ...         80.6   
3         213.310     15.334  ...         80.6   
4         213.310     15.334  ...         80.6   
...           ...        ...  ...          ...   
13818     155.505      7.225  ...         65.4   
13819         NaN        NaN  ...          NaN   
13820     633.527     16.620  ...         94.4   
13821         NaN        NaN  ...          NaN   
13822         NaN        NaN  ...          NaN   

                                                geometry          prov_key  \
0      LINESTRING (121.34927 15.99296, 121.34949 15.9...      NUEVAVIZCAYA   
1      LINESTRING (125.41293 7.06458, 125.41301 7.064...       DAVAODELSUR   
2      LINESTRING (125.51876 7.56097, 125.51912 7.560...       DAVAODELSUR   
3      LINESTRING (125.51876 7.56097, 125.51912 7.560...       DAVAODELSUR   
4      LINESTRING (125.51876 7.56097, 125.51912 7.560...     DAVAODELNORTE   
...                                                  ...               ...   
13818  MULTILINESTRING ((121.17283 14.54152, 121.1724...             RIZAL   
13819  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  ZAMBOANGASIBUGAY   
13820  LINESTRING (124.24519 8.25548, 124.24604 8.255...     LANAODELNORTE   
13821  LINESTRING (125.23851 12.41278, 125.23868 12.4...     NORTHERNSAMAR   
13822  LINESTRING (125.23851 12.41278, 125.23868 12.4...     NORTHERNSAMAR   

       Runoff_from_PNHMD  Runoff_mm_yr     Q_m3s  Power_MW   HPI Final_Status  \
0                  9.570        9570.0  0.008401  0.177462  61.1         Good   
1                  5.330        5330.0       NaN       NaN  30.0         Poor   
2                  5.330        5330.0  0.000535  0.006020  60.0         Good   
3                  5.330        5330.0  0.000535  0.006020  60.0         Good   
4               3740.060     3740060.0  0.375704  4.224545  89.1    Excellent   
...                  ...           ...       ...       ...   ...          ...   
13818              0.870         870.0  0.000109  0.001732  43.1

### Final Dataset (God Willing)

In [ ]:
# Rename each column for readability 
nice_cols = {
    "ADM1_EN": "Region",
    "ADM2_EN_prov": "Province",
    "ADM3_EN": "City_Muni",
    "Catch_area": "Catchment_km2",
    "Catch_rel": "Relief_m",
    "Runoff_mm_yr": "Runoff_mm_per_yr",
    "Q_m3s": "Flow_m3s",
    "Power_MW": "Power_MW",
    "HPI": "HPI",
    "Final_Status": "Status",
    "Final_Usable": "Usable",
}
order = [
    "name","waterway","Region","Province","City_Muni","Catchment",
    "Catchment_km2","Relief_m","Runoff_mm_per_yr","Flow_m3s","Power_MW","HPI","Status","Usable","geometry"
]
final_river = (
    scored_river.rename(columns=nice_cols)
                .loc[:, [c for c in order if c in scored_river.rename(columns=nice_cols).columns]]
                .copy()
)

# Formatting 
final_river["Catchment_km2"]     = final_river["Catchment_km2"].round(1)
final_river["Relief_m"]          = final_river["Relief_m"].round(0)
final_river["Runoff_mm_per_yr"]  = final_river["Runoff_mm_per_yr"].round(0)
final_river["Flow_m3s"]          = final_river["Flow_m3s"].round(4)   # small flows can matter
final_river["Power_MW"]          = final_river["Power_MW"].round(4)
final_river["HPI"]               = final_river["HPI"].round(1)


In [20]:
final_river

name waterway                           Region  \
0          Cagayan River    river       Region II (Cagayan Valley)   
1            Lopoy Creek   stream         Region XI (Davao Region)   
2          Tuganay River    river         Region XI (Davao Region)   
3          Tuganay River    river         Region XI (Davao Region)   
4          Tuganay River    river         Region XI (Davao Region)   
...                  ...      ...                              ...   
13818       Angono River    river         Region IV-A (Calabarzon)   
13819  Pamoantogbo River    river  Region IX (Zamboanga Peninsula)   
13820     Mandulog River    river     Region X (Northern Mindanao)   
13821        Gamay River    river    Region VIII (Eastern Visayas)   
13822        Gamay River    river    Region VIII (Eastern Visayas)   

                Province          City_Muni       Catchment  Catchment_km2  \
0          Nueva Vizcaya  Alfonso Castaneda         Cagayan        27684.1   
1          Davao del Sur         Davao City             NaN            NaN   
2          Davao del Sur         Davao City           Tagum         3167.9   
3          Davao del Sur        Santo Tomas           Tagum         3167.9   
4        Davao del Norte         Davao City           Tagum         3167.9   
...                  ...                ...             ...            ...   
13818              Rizal             Angono  Marikina_Pasig         3961.0   
13819  Zamboanga Sibugay              Payao             NaN            NaN   
13820    Lanao del Norte        Iligan City        Mandulog          776.0   
13821     Northern Samar              Gamay             NaN            NaN   
13822     Northern Samar            Mapanas             NaN            NaN   

       Relief_m  Runoff_mm_per_yr  Flow_m3s  Power_MW   HPI     Status  \
0        2871.0            9570.0    0.0084    0.1775  61.1       Good   
1           NaN            5330.0       NaN       NaN  30.0       Poor   
2        1528.0            5330.0    0.0005    0.0060  60.0       Good   
3        1528.0            5330.0    0.0005    0.0060  60.0       Good   
4        1528.0         3740060.0    0.3757    4.2245  89.1  Excellent   
...         ...               ...       ...       ...   ...        ...   
13818    2155.0             870.0    0.0001    0.0017  43.1   Moderate   
13819       NaN            7730.0       NaN       NaN  30.0       Poor   
13820    1400.0           12050.0    0.0003    0.0031  60.0       Good   
13821       NaN           77414.0       NaN       NaN  30.0       Poor   
13822       NaN           77414.0       NaN       NaN  30.0       Poor   

       Usable                                           geometry  
0        True  LINESTRING (121.34927 15.99296, 121.34949 15.9...  
1       False  LINESTRING (125.41293 7.06458, 125.41301 7.064...  
2        True  LINESTRING (125.51876 7.56097, 125.51912 7.560...  
3        True  LINESTRING (125.51876 7.56097, 125.51912 7.560...  
4        True  LINESTRING (125.51876 7.56097, 125.51912 7.560...  
...       ...                                                ...  
13818   False  MULTILINESTRING ((121.17283 14.54152, 121.1724...  
13819   False  MULTILINESTRING ((122.8461 7.53302, 122.8462 7...  
13820    True  LINESTRING (124.24519 8.25548, 124.24604 8.255...  
13821   False  LINESTRING (125.23851 12.41278, 125.23868 12.4...  
13822   False  LINESTRING (125.23851 12.41278, 125.23868 12.4...  

[13823 rows x 15 columns]

In [21]:
#def export_hydro_to_csv(gdf, out_path="river_sites.csv"):
#    df = gdf.copy()
#    df["geometry_wkt"] = df.geometry.to_wkt()
#    df = pd.DataFrame(df.drop(columns=["geometry"]))
#    
#    # convert data frame to CSV 
#    df.to_csv(out_path, index=False, encoding="utf-8")
#    print(f"CSV saved successfully → {out_path}")



In [22]:
#export_hydro_to_csv(final_river, "River_dataset.csv")